op-challenger/runner/runner.go

```diff
@@ -115,20 +108,47 @@ func (r *Runner) loop(ctx context.Context, traceType types.TraceType, client *so
 	}
 }
 
-func (r *Runner) runOnce(ctx context.Context, traceType types.TraceType, client *sources.RollupClient, caller *batching.MultiCaller) error {
+func (r *Runner) runAndRecordOnce(ctx context.Context, traceType types.TraceType, client *sources.RollupClient, caller *batching.MultiCaller) {
+	recordError := func(err error, traceType types.TraceType, m Metricer, log log.Logger) {
+		if errors.Is(err, ErrUnexpectedStatusCode) {
+			log.Error("Incorrect status code", "type", traceType, "err", err)
+			m.RecordInvalid(traceType)
+		} else if err != nil {
+			log.Error("Failed to run", "type", traceType, "err", err)
+			m.RecordFailure(traceType)
+		} else {
+			log.Info("Successfully verified output root", "type", traceType)
+			m.RecordSuccess(traceType)
+		}
+	}
+
 	prestateHash, err := r.getPrestateHash(ctx, traceType, caller)
 	if err != nil {
-		return err
+		recordError(err, traceType, r.m, r.log)
+		return
 	}
 
 	localInputs, err := r.createGameInputs(ctx, client)
 	if err != nil {
-		return err
+		recordError(err, traceType, r.m, r.log)
+		return
 	}
 	dir, err := r.prepDatadir(traceType)
 	if err != nil {
-		return err
+		recordError(err, traceType, r.m, r.log)
+		return
+	}
+	err = r.runOnce(ctx, traceType, prestateHash, localInputs, dir)
+	recordError(err, traceType, r.m, r.log)
+
+	if r.addMTCannonPrestate != (common.Hash{}) {
+		// reuse the trace directory
+		err := r.runMTCannonOnce(ctx, localInputs, dir)
+		recordError(err, traceType, r.m, r.log.With("mt-cannon", true))
 	}
```

Comment:
It would be better if we run the MT cannon in parallel. Each trace type executes at the same time in vm-runner and there should be enough compute capacity to run two cannon's in parallel.

Reply:
Will do. In the case we run MT-Cannon with Cannon, we will end up fetching the same preimages twice. But that's not too bad.

Reply:
Wouldn't they be different preimages (one single threaded and one multithreaded)?

Reply:
Yeah. I mean that the preimage keys would be mostly identical. So one could ideally reuse the single-threaded preimages for multi-threaded (ignoring op-geth shenanigans with multithreaded block finalization).

Reply:
Oh preimages not prestate sorry. Yes that's true, but a worthwhile trade off to effectively halve the execution time.